// @strapi/content-manager/dist/admin/preview/components/InputPopover.mjs
import { jsxs, Fragment, jsx } from 'react/jsx-runtime';
import 'react';
import { createContext } from '@strapi/admin/strapi-admin';
import { Box, Popover } from '@strapi/design-system';
import { InputRenderer as MemoizedInputRenderer } from '../../pages/EditView/components/InputRenderer.mjs';
import { usePreviewContext } from '../pages/Preview.mjs';

const [InputPopoverProvider, useInputPopoverContext] = createContext('InputPopover');
const InputPopover = ({ documentResponse })=>{
    const iframeRef = usePreviewContext('VisualEditingPopover', (state)=>state.iframeRef);
    const popoverField = usePreviewContext('VisualEditingPopover', (state)=>state.popoverField);
    const setPopoverField = usePreviewContext('VisualEditingPopover', (state)=>state.setPopoverField);
    if (!popoverField || !documentResponse.schema || !iframeRef.current) {
        return null;
    }
    const iframeRect = iframeRef.current.getBoundingClientRect();
    return /*#__PURE__*/ jsxs(Fragment, {
        children: [
            /*#__PURE__*/ jsx(Box, {
                position: 'fixed',
                top: iframeRect.top + 'px',
                left: iframeRect.left + 'px',
                width: iframeRect.width + 'px',
                height: iframeRect.height + 'px',
                zIndex: 4
            }),
            /*#__PURE__*/ jsx(InputPopoverProvider, {
                children: /*#__PURE__*/ jsxs(Popover.Root, {
                    open: true,
                    onOpenChange: (open)=>!open && setPopoverField(null),
                    children: [
                        /*#__PURE__*/ jsx(Popover.Trigger, {
                            children: /*#__PURE__*/ jsx(Box, {
                                position: "fixed",
                                width: popoverField.position.width + 'px',
                                height: popoverField.position.height + 'px',
                                top: 0,
                                left: 0,
                                transform: `translate(${iframeRect.left + popoverField.position.left}px, ${iframeRect.top + popoverField.position.top}px)`
                            })
                        }),
                        /*#__PURE__*/ jsx(Popover.Content, {
                            sideOffset: 4,
                            children: /*#__PURE__*/ jsx(Box, {
                                padding: 4,
                                width: "400px",
                                children: /*#__PURE__*/ jsx(MemoizedInputRenderer, {
                                    document: documentResponse,
                                    attribute: documentResponse.schema.attributes[popoverField.path],
                                    label: popoverField.path,
                                    name: popoverField.path,
                                    type: documentResponse.schema.attributes[popoverField.path].type,
                                    visible: true
                                })
                            })
                        })
                    ]
                })
            })
        ]
    });
};
function useHasInputPopoverParent() {
    const context = useInputPopoverContext('useHasInputPopoverParent', ()=>true, false);
    // useContext will return undefined if the called is not wrapped in the provider
    return context !== undefined;
}

export { InputPopover, useHasInputPopoverParent };
