// @_sh/strapi-plugin-ckeditor/dist/server/index.js
"use strict";
const name = "@_sh/strapi-plugin-ckeditor";
const version = "6.0.3";
const description = "Integrates CKEditor 5 into your Strapi project as a fully customizable custom field. (Community Edition)";
const keywords = [
  "strapi",
  "ckeditor",
  "ckeditor5",
  "ckeditor 5",
  "wysiwyg",
  "rich text",
  "editor"
];
const author = {
  name: "nshenderov"
};
const homepage = "https://market.strapi.io/plugins/@_sh-strapi-plugin-ckeditor";
const repository = {
  type: "git",
  url: "https://github.com/nshenderov/strapi-plugin-ckeditor.git"
};
const strapi = {
  name: "ckeditor5",
  displayName: "CKEditor 5",
  description: "Integrates CKEditor 5 into your Strapi project as a fully customizable custom field. (Community Edition)",
  kind: "plugin"
};
const type = "commonjs";
const exports$1 = {
  "./package.json": "./package.json",
  ".": {
    types: "./dist/admin/src/index.d.ts",
    source: "./admin/src/index.ts",
    "import": "./dist/admin/index.mjs",
    require: "./dist/admin/index.js",
    "default": "./dist/admin/index.js"
  },
  "./strapi-admin": {
    types: "./dist/admin/src/index.d.ts",
    source: "./admin/src/index.ts",
    "import": "./dist/admin/index.mjs",
    require: "./dist/admin/index.js",
    "default": "./dist/admin/index.js"
  },
  "./strapi-server": {
    types: "./dist/server/src/index.d.ts",
    source: "./server/src/index.ts",
    "import": "./dist/server/index.mjs",
    require: "./dist/server/index.js",
    "default": "./dist/server/index.js"
  }
};
const files = [
  "dist"
];
const scripts = {
  build: "strapi-plugin build",
  watch: "strapi-plugin watch",
  "watch:link": "strapi-plugin watch:link",
  verify: "strapi-plugin verify",
  "test:ts:front": "run -T tsc -p admin/tsconfig.json",
  "test:ts:back": "run -T tsc -p server/tsconfig.json",
  release: "release-it",
  "release:info": "release-it --changelog"
};
const dependencies = {
  "@ckeditor/ckeditor5-react": "~9.5.0",
  "@strapi/design-system": "2.0.0-rc.18",
  "@strapi/icons": "2.0.0-rc.18",
  ckeditor5: "~45.2.0",
  lodash: "4.17.21",
  "sanitize-html": "2.13.0",
  yup: "0.32.9"
};
const devDependencies = {
  "@release-it/conventional-changelog": "10.0.0",
  "@strapi/sdk-plugin": "5.3.2",
  "@strapi/strapi": "5.11.0",
  "@strapi/typescript-utils": "5.11.0",
  "@types/react": "^18.3.12",
  "@types/react-dom": "^18.3.1",
  "@types/sanitize-html": "2.13.0",
  "@typescript-eslint/eslint-plugin": "7.0.0",
  "@typescript-eslint/parser": "7.0.0",
  eslint: "8.2.0",
  "eslint-config-airbnb": "19.0.4",
  "eslint-config-airbnb-typescript": "18.0.0",
  "eslint-config-prettier": "9.1.0",
  "eslint-plugin-import": "2.25.3",
  "eslint-plugin-jsx-a11y": "6.10.2",
  "eslint-plugin-prettier": "5.2.1",
  "eslint-plugin-react": "7.28.0",
  "eslint-plugin-react-hooks": "4.3.0",
  prettier: "3.4.0",
  react: "18.3.1",
  "react-dom": "18.3.1",
  "react-router-dom": "6.30.0",
  "release-it": "18.1.2",
  "styled-components": "6.1.15",
  typescript: "5.6.3"
};
const peerDependencies = {
  "@strapi/strapi": "^5.0.0",
  react: "^18.0.0",
  "react-dom": "^18.0.0",
  "react-router-dom": "^6.0.0",
  "styled-components": "^6.0.0"
};
const engines = {
  node: ">=18.0.0 <=22.x.x",
  npm: ">=6.0.0"
};
const license = "MIT";
const pluginPkg = {
  name,
  version,
  description,
  keywords,
  author,
  homepage,
  repository,
  strapi,
  type,
  exports: exports$1,
  files,
  scripts,
  dependencies,
  devDependencies,
  peerDependencies,
  engines,
  license
};
const index = {
  register: ({ strapi: strapi2 }) => {
    strapi2.customFields.register({
      name: "CKEditor",
      plugin: pluginPkg.strapi.name,
      type: "richtext"
    });
  }
};
module.exports = index;
